Assets/Mirror/Runtime/NetworkWriter.cs

```diff
@@ -19,28 +20,70 @@ public class NetworkWriter
 
         // 'int' is the best type for .Position. 'short' is too small if we send >32kb which would result in negative .Position
         // -> converting long to int is fine until 2GB of data (MAX_INT), so we don't have to worry about overflows here
-        public int Position;
-
+        int position;
         int length;
 
-        public int Length
+        public int Length => length;
+
+        public int Position
         {
-            get => length;
-            private set
+            get => position;
+            set
             {
-                EnsureCapacity(value);
-                length = value;
-                if (Position > length)
-                    Position = length;
+                position = value;
+                EnsureLength(value);
+            }
+        }
+
+        /// <summary>
+        /// Reset both the position and length of the stream
+        /// </summary>
+        /// <remarks>
+        /// Leaves the capacity the same so that we can reuse this writer without extra allocations
+        /// </remarks>
+        public void Reset()
+        {
+            position = 0;
+            length = 0;
+        }
+
+        /// <summary>
+        /// Sets length, moves position if it is greater than new length
+        /// </summary>
+        /// <param name="newLength"></param>
+        /// <remarks>
+        /// Zeros out any extra length created by setlength
+        /// </remarks>
+        public void SetLength(int newLength)
+        {
+            int oldLength = length;
+
+            // use EnsureCapacity 
+            EnsureLength(newLength);
+
+            // zero out new length
+            if (oldLength < newLength)
+            {
+                Array.Clear(buffer, oldLength, newLength - oldLength);
             }
+
+            length = newLength;
+            position = Mathf.Min(position, length);
         }
 
-        void EnsureCapacity(int value)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
does inlining make a big difference?
and can we assume that this works on all platforms and doesn't cause us extra issues?

Reply:
This is what I know about it, from 2017 https://youtu.be/_wxitgdx-UI?t=1118

It might not work on all platforms, but it shouldn't cause any issues for us.

The function is called in 6 places so it is probably not worth manually Inlining